javascript/ql/src/semmle/javascript/frameworks/Vue.qll

```diff
@@ -0,0 +1,359 @@
+    or
+    result = DataFlow::moduleImport("vue")
+  }
+
+  /**
+   * A call to `vue.extend`.
+   */
+  private class VueExtend extends DataFlow::CallNode {
+    VueExtend() { this = vue().getAMemberCall("extend") }
+  }
+
+  private newtype TInstance =
+    MkVueInstance(DataFlow::NewNode def) { def = vue().getAnInstantiation() } or
+    MkExtendedVue(VueExtend extend) or
+    MkExtendedInstance(VueExtend extend, DataFlow::NewNode sub) {
+      sub = extend.getAnInstantiation()
+    } or
+    MkComponent(DataFlow::CallNode def) { def = vue().getAMemberCall("component") } or
+    MkSingleFileComponent(VueFile file)
+
+  /**
+   * A Vue instance definition.
+   *
+   * This includes both explicit instantiations of Vue objects, and
+   * implicit instantiations in the form of components or Vue
+   * extensions that have not yet been instantiated to a Vue instance.
+   *
+   * The following instances are recognized:
+   *   - `new Vue({...})`
+   *   - `Vue.extend({...})`
+   *   - `new ExtendedVue({...})`
+   *   - `Vue.component("my-component", {...})`
+   *   - single file components in .vue files
+   */
+  abstract class Instance extends TInstance {
+    /** Gets a textual representation of this element. */
+    abstract string toString();
+
+    /**
+     * Holds if this element is at the specified location.
+     * The location spans column `startcolumn` of line `startline` to
+     * column `endcolumn` of line `endline` in file `filepath`.
+     * For more information, see
+     * [LGTM locations](https://lgtm.com/help/ql/locations).
+     */
+    predicate hasLocationInfo(
+      string filepath, int startline, int startcolumn, int endline, int endcolumn
+    ) {
+      filepath = "" and
+      startline = 0 and
+      startcolumn = 0 and
+      endline = 0 and
+      endcolumn = 0
+    }
+
+    /**
+     * Gets the node for option `name` for this instance, this does not include
+     * those from extended objects and mixins.
+     */
+    abstract DataFlow::Node getOwnOption(string name);
+
+    /**
+     * Gets the node for option `name` for this instance, including those from
+     * extended objects and mixins.
+     */
+    DataFlow::Node getOption(string name) {
+      result = getOwnOption(name)
+      or
+      exists(DataFlow::SourceNode extendsVal | extendsVal.flowsTo(getOwnOption("extends")) |
+        result = extendsVal.(DataFlow::ObjectLiteralNode).getAPropertyWrite(name).getRhs()
+        or
+        exists(ExtendedVue extend |
+          MkExtendedVue(extendsVal) = extend and
+          result = extend.getOption(name)
+        )
+      )
+      or
+      exists(DataFlow::ArrayCreationNode mixins, DataFlow::ObjectLiteralNode mixin |
+        mixins.flowsTo(getOwnOption("mixins")) and
+        mixin.flowsTo(mixins.getAnElement()) and
+        result = mixin.getAPropertyWrite(name).getRhs()
+      )
+    }
+
+    /**
+     * Gets the node for the `data` option object of this instance.
+     */
+    DataFlow::Node getData() {
```

Comment:
It looks like the result of this predicate is always a `SourceNode`; maybe we want to make that explicit in the result type? Also, why does this predicate return local sources, while `getOption` above does not?

Reply:
Fixed, clients need to use `flowsTo` now. 
One of the branches in `getAnAccessor`/`getGetter` also had this problem.